Core/Frameworks/Baikal/Core/LDAP.php

```diff
@@ -0,0 +1,319 @@
+     *
+     * @return bool
+     */
+    protected function doesBind(&$conn, $dn, $password) {
+        try {
+            $bind = ldap_bind($conn, $dn, $password);
+            if ($bind) {
+                return true;
+            }
+        } catch (\ErrorException $e) {
+            error_log($e->getMessage());
+            error_log(ldap_error($conn));
+        }
+
+        return false;
+    }
+
+    /**
+     * Creates the backend object.
+     *
+     * @param string $ldap_uri
+     * @param string $ldap_dn
+     * @param string $ldap_cn
+     * @param string $ldap_mail
+     */
+    public function __construct(\PDO $pdo, $table_name = 'users', $ldap_mode = 'DN', $ldap_uri = 'ldap://127.0.0.1', $ldap_bind_dn = 'cn=baikal,ou=apps,dc=example,dc=com', $ldap_bind_password = '', $ldap_dn = 'mail=%u', $ldap_cn = 'cn', $ldap_mail = 'mail', $ldap_search_base = 'ou=users,dc=example,dc=com', $ldap_search_attribute = 'uid=%U', $ldap_search_filter = '(objectClass=*)', $ldap_group = 'cn=baikal,ou=groups,dc=example,dc=com') {
+        $this->pdo                   = $pdo;
+        $this->table_name            = $table_name;
+        $this->ldap_mode             = $ldap_mode;
+        $this->ldap_uri              = $ldap_uri;
+        $this->ldap_bind_dn          = $ldap_bind_dn;
+        $this->ldap_bind_password    = $ldap_bind_password;
+        $this->ldap_dn               = $ldap_dn;
+        $this->ldap_cn               = $ldap_cn;
+        $this->ldap_mail             = $ldap_mail;
+        $this->ldap_search_base      = $ldap_search_base;
+        $this->ldap_search_attribute = $ldap_search_attribute;
+        $this->ldap_search_filter    = $ldap_search_filter;
+        $this->ldap_group            = $ldap_group;
+    }
+
+    /**
+     * Connects to an LDAP server and tries to authenticate.
+     *
+     * @param string $username
+     * @param string $password
+     *
+     * @return bool
+     */
+    protected function ldapOpen($username, $password) {
+        $conn = ldap_connect($this->ldap_uri);
+        if (!$conn) {
+            return false;
+        }
+        if (!ldap_set_option($conn, LDAP_OPT_PROTOCOL_VERSION, 3)) {
+            return false;
+        }
+
+        $success = false;
+
+        if ($this->ldap_mode == 'DN') {
+            $this->patternReplace($dn, $username);
+
+            $success = $this->doesBind($conn, $dn, $password);
+        } elseif ($this->ldap_mode == 'Attribute' || $this->ldap_mode == 'Group') {
+            try {
+                if (!$this->doesBind($conn, $this->ldap_bind_dn, $this->ldap_bind_password)) {
+                    return false;
+                }
+
+                $attribute = $this->ldap_search_attribute;
+                $this->patternReplace($attribute, $username);
+
+                $result = ldap_get_entries($conn, ldap_search($conn, $this->ldap_search_base, '(' . $attribute . ')', [explode('=', $attribute, 2)[0]], 0, 1, 0, LDAP_DEREF_ALWAYS, []))[0];
```

Comment:
This line is very long. Consider adding 1-2 line breaks (where logical) to make it easier to read. We don't have a strict line limit check in place, but please aim for at most ~120 characters.

Reply:
I personally don't break lines of code, I actually think they worsen readability.
Still, you want it? You got it.